contracts/StorageManager.sol

```diff
@@ -172,177 +156,200 @@ contract StorageManager {
+        - if Agreement is terminated Consumer can withdraw all remaining funds
+    @param dataReference the data reference of agreement to be funds withdrawn from
     @param provider the address of the provider of the Offer.
     */
-    function terminateAgreement(bytes32[] memory dataReference, address provider) public payable {
-        bytes32 agreementReference = getAgreementReference(dataReference);
-        Agreement storage agreement = offerRegistry[provider].agreementRegistry[agreementReference];
-        // NO_OVERFLOW reasoning: startDate is always less than now. agreement.billingPeriod is verified not to be 0 in function: newRequest
-        uint256 periodsPast = ((now - agreement.startDate) / agreement.billingPeriod) + 1;
-        require(
-            agreement.numberOfPeriodsWithdrawn + periodsPast <
-            agreement.numberOfPeriodsDeposited,
-            "StorageManager: agreement expired or in last period"
-        );
-        agreement.numberOfPeriodsDeposited = 0;
-        agreement.numberOfPeriodsWithdrawn = 0;
-        //NO_OVERFLOW reasoning: REF_CAPACITY
-        offerRegistry[msg.sender].availableCapacity = offerRegistry[msg.sender].availableCapacity + agreement.size;
-        agreement.startDate = 0;
-        //NO_OVERFLOW reasoning:  verified in "agreement expired" require and see REF_MAX_TRANSFER
-        msg.sender.transfer(agreement.numberOfPeriodsDeposited - periodsPast * agreement.billingPrice);
-        emit AgreementStopped(agreementReference);
+    function withdrawFunds(bytes32[] memory dataReference, address provider, uint256 amount) public payable {
+        Offer storage offer = offerRegistry[provider];
+        require(offer.availableCapacity != 0, "StorageManager: Offer for this Provider doesn't exist");
+
+        bytes32 agreementReference = getAgreementReference(dataReference, msg.sender);
+        Agreement storage agreement = offer.agreementRegistry[agreementReference];
+        require(agreement.size != 0, "StorageManager: Agreement for this Offer doesn't exist");
+
+        uint256 maxWithdrawableFunds;
+        if (agreement.lastPayoutDate == 0) {
+            // Agreement is inactive, consumer can withdraw all funds
+            maxWithdrawableFunds = agreement.availableFunds;
+        } else {
+            // Consumer can withdraw all funds except for those already used for past storage hosting
+            // AND for current period
+            maxWithdrawableFunds = agreement.availableFunds - _calculateSpentFunds(agreement) - (agreement.billingPrice * agreement.size);
+        }
+
+        if (amount == 0) {
+            amount = maxWithdrawableFunds;
+        }
+
+        require(amount <= maxWithdrawableFunds, "StorageManager: Amount is too big");
+        agreement.availableFunds = agreement.availableFunds.sub(amount);
+        msg.sender.transfer(amount);
```

Comment:
What about check if amount > 0? It is only to caller spent less gas...
require (amount > 0, "nothing to withdraw");

Suggestion to not use transfer because https://eips.ethereum.org/EIPS/eip-1884
(bool success, ) = msg.sender.call.value(amount)("");
require(success, "Transfer failed.");

Reply:
Thanks about the `transfer` info.